extract `propertyDependencies` (#1450)
json-schema-org/json-schema-spec: proposals/propertyDependencies.md

# JSON Schema Proposal: Add `propertyDependencies` Keyword

## Abstract

The `propertyDependencies` keyword is a more friendly way to select between two
or more schemas to validate an instance against than is currently supported by
JSON Schema.

## Status

> **Current Status**: PROPOSAL

Reviewer: See, https://github.com/json-schema-org/json-schema-spec/pull/1450#discussion_r1362772967

TODO: We should have a short standard blurb outlining the stages involved in a
feature making its way to stable status.

TODO: Link to a document that describes the proposal => stable process in

> detail.

Author: Similar here.  I think just a link to the statuses (we should decide on stages/statuses at some point) document would suffice.

Reviewer: I think it depends on how public-facing we expect this document to be. If the audience is mainly internal, then a link would be sufficient. If it's intended to be more public, I think it would be beneficial to have a short canned blurb about what the status means so causal readers can get the gist of what the status means without having to follow the link to a detailed document about our process.

Reviewer 2: > I think it depends on how public-facing we expect this document to be.

Disagree here. We should do what the simplest is for us in this case. My feeling is that these documents are going to be pretty meaningless without a reasoable understanding of the process and where such a document currently sits within that.

    }
  ]
}
```

This one is a variation of `if` that combines `if`, `properties`, and `required`
to reduce boilerplate. It's also essentially a variation of the previous example
with better names. This avoids to error proneness problem, but it's still too
verbose.

```json
{
  "allOf": [
    {
      "ifProperties": {
        "foo": { "const": "aaa" }
      },
      "then": { "$ref": "#/$defs/foo-aaa" }
    },
    {
      "ifProperties": {
        "foo": { "const": "bbb" }
      },
      "then": { "$ref": "#/$defs/foo-aaa" }
    }
  ]
}
```

All of the previous alternatives use a schema as the discriminator. This
alternative is a little less powerful in that it can only match on exact values,
but it successfully addresses the problems we're concerned about with the
current approaches. The only issue with this alternative is that it's not as
intuitive as the chosen solution.

```json
{
  "propertyDepenencies": {
    "foo": [
      ["aaa", { "$ref": "#/$defs/foo-aaa" }],
      ["bbb", { "$ref": "#/$defs/foo-bbb" }]
    ]
  }
}
```

## Specification

<!-- TODO: Update to where this actually ends up getting hosted. This is a guess
for now. -->

> <https://json-schema.org/specification/extensions/propertyDependencies.html>

Author: But here you're linking to a published doc.  Seems like publishing this doc would help build a case for the feature's inclusion.

Reviewer: We were talking about this as an ADR type of document. Usually that kind of thing has an internal audience, so I was expecting that this would probably just be on GitHub. I'm not opposed to thinking about putting it on the website for more exposure, but I wouldn't put it up in the form of a spec document like our main specs or extension specs. What I had in mind, was that part of the proposal making it to Stage-1 (or whatever we end up calling it) was adding a UJS entry describing the keyword. That would expose users to proposed keywords that they might want to use and we can link to the proposal and the spec for people who want to dig deeper.

Author: Publishing it leans into my previous point about a dev needing to convince their mgr/superior to use an experimental feature.  I think this content is useful.  Another reason I prefer the single doc approach.

Your concern before was that a dev wouldn't care about the "fluff" but would want to know about the stuff at the bottom and they shouldn't have to scroll all the way down there:

> the implementer needs to scroll past a bunch of stuff they don't care about in order to get to the information they need

They'll get a TOC that they can use to go straight to what they need.

Reviewer: > Publishing it leans into my previous point about a dev needing to convince their mgr/superior to use an experimental feature.

Good point. That's a reason to consider having the audience be more than just internal. However, I'm not sure that necessarily means it should be on the website. The document is still public on GitHub.

> I think this content is useful for consumers of JSON Schema.

I think having something specifically for consumers and authors in UJS would be more useful because it's specifically focused for that audience.

> Another reason I prefer the single doc approach.

I'm not seeing the connection here. I think the separate documents approach makes sense in this case. The manager cares about what the trade-offs are. They can get that just from the proposal document. The consumer/author cares about what the feature does and how to use it. They can get that from just the UJS documentation. The implementer cares about what they need to build. They can get that from just the spec. Of course there's going to be some overlap and people wearing multiple hats, but I think the separation of concerns by audience makes a lot of sense.

> They'll get a TOC that they can use to go straight to what they need.

I take your point, but that doesn't address the other concerns I brought up. An extension could have a life outside of the proposal process. It might start off as an extension spec that was later proposed to be added to the spec. It might have been a proposal that was rejected, but still wants to live on as a third-party extension. Decoupling the spec from the proposal process allows for these cases.

# JSON Schema: The `propertyDependencies` Keyword

## Abstract

The `propertyDependencies` keyword is a more friendly way to select between two
or more schemas to validate an instance against than is currently supported by
JSON Schema.

## Status

**Current Status**: PROPOSAL

TODO: We should have a short standard blurb outlining the stages involved in a
feature making its way to stable status.

TODO: Link to a document that describes the proposal => stable process in
detail.

## Note to Readers

The issues list for this document can be found at
<https://github.com/json-schema-org/json-schema-spec/issues?q=is%3Aissue+propertydependencies>.

For additional information, see <https://json-schema.org/>.

To provide feedback, use this issue tracker or any of the communication methods
listed on the homepage.

## Table of Contents

## Conventions and Terminology

All conventions and terms used and defined by the [JSON Schema Core
specification](../jsonschema-core.html) also apply to this document.

## Overview

> ### Problem Statement
> A common need in JSON Schema is to select between one schema or another to

Author: ```suggestion
### Problem Statement

A common need in JSON Schema is to select between one schema or another to
```

validate an instance based on the value of some property in the JSON instance.
There are a several patterns people use to accomplish this, but they all have
significant [problems](#problems).

OpenAPI solves this problem with the `discriminator` keyword. However, their
approach is more oriented toward code generation concerns, is poorly specified

> when it comes to validation, and is couple to OpenAPI concepts that don't exist

Author: ```suggestion
when it comes to validation, and is coupled to OpenAPI concepts that don't exist
```

is JSON Schema. Therefore, it's necessary to define something new rather than
adopt `discriminator`.

### Solution

The `dependentSchemas` keyword is very close to what is needed except it checks
for the presence of a property rather than it's value. The chosen solution is to
build on that concept to solve this problem.

```jsonschema
{
  "propertyDependencies": {
    "foo": {
      "aaa": { "$ref": "#/$defs/foo-aaa" }
    }
  }
}
```

> The validation result is equivalent to the following schema.

Author: Is it correct to say that the _schemas_ are equivalent?  It seems strange to isolate that the validation result is equivalent.

Reviewer: Good point. I meant "validation _behavior_". I specifically wanted to call out validation behavior to exclude annotation behavior. There are multiple schemas that have equivalent validation behavior to the is keyword but not necessarily equivalent annotation behavior. I didn't want anyone to assume that this example says anything about annotation behavior of `propertyDependencies`.

```jsonschema
{
  "if": {
    "properties": {
      "foo": { "const": "aaa" }
    },
    "required": ["foo"]
  },
  "then": { "$ref": "#/$defs/foo-aaa" }
}
```

### Limitations

> The problem of choosing an alternative based on a property value could apply for

Author: Maybe choose a different word from "alternative" here since it has special meaning in the context of this document.

```suggestion
The problem of choosing a subschema based on a property value could apply for
```

Reviewer: Good call. I didn't notice that.

a value of any JSON type, but `propertyDependencies` only solves this problem
when the value is a string. One of the main goals of this keyword is to define
something that's intuitive enough and easy enough to use that people will

> actually use it rather than fallback to `oneOf` because it's simple. Achieving
> those goals means that some trade-offs need to be made. {{alternatives}} lists
> some alternatives that we considered.

Author: I don't think these last couple of sentences are needed.  I think trade-offs are implied by having a "limitations" section.

```suggestion
actually use it rather than fallback to `oneOf` because it's simple.
```

Reviewer: The purpose of those words is to say that none of the alternatives we considered satisfy our goals for the keyword. I think that's important to say. It also provides a link for readers who want to do a deep dive on what alternatives we considered. I think that's important too. Maybe those sentences could use some wordsmithing, but I think they're both saying important things.

> ## A Vocabulary for Applying Subschemas

Reviewer: Sorry, I forgot to explain why I changed that. This was a combination of a few things. I thought it was unnecessary for the "Proposal" content to be empty. The "Target for Change" content makes perfect sense in that space and we don't need the extra section.

After moving that content, my next thought was that the content was fragile. It's referencing a section number and its title in the spec. Either of those could change and that reference would be wrong. I wanted to both reduce that fragility and make it more explicit that the change is to add this keyword to a specific vocabulary. So, I changed it to reference the vocabulary URI the keyword is being added to. In theory that should be permanent and never change. To be helpful, I also linked to the section in the spec that describes the vocabulary. I added named anchor to use for this reference specifically to make it less fragile.

However, later I was trying to replace the `Proposal` heading with something that doesn't use the word "proposal". I couldn't think of anything at the time, but thought it would be interesting mimic the hierarchy of the spec. (Interestingly, that's also what they did [here](https://tc39.es/proposal-export-default-from/).) Of course, this goes against the anti-fragile goal I had for the previous change, but I had no other ideas at the time. What I probably should have done was bring back the "Target for Change" I had removed from before and used that heading instead. However, if you're ok with the word "Extension", I think I like that better. Let me know which you prefer and I'll drop the spec-matched hierarchy. I wasn't happy with the fragility of that anyway.

This document adds the `propertyDependencies` keyword to the
`https://json-schema.org/vocab/applicator` [applicator

> vocabulary](../jsonschema-core.html#applicatorvocab).

Reviewer: I think this is another case of us having different ideas about what this document is for. I'm thinking that this document should be able to stand on its own as an extension to JSON Schema whether it gets accepted into the spec or not is a separate concern. It's use/implementation isn't dependent on being accepted to the spec, so I'd like to have some decoupling in that respect. It seems plenty to say its name and what vocabulary its part of and then when/if it's integrated into the spec, where exactly it ends up is the concern of the document it's being integrated into.

If you like the separate documents idea, I can see information about where exactly the change is expected to go to be in the "proposal" document, but honestly, I don't think it's a concern until we're actually at the stage of putting it into the spec.

no way to specify what that value is and therefore needs to evaluate the entire
schema. This is made worse in that every alternative needs to be fully validated
to ensure that only one of the alternative passes and all the others fail. This
last problem can be avoided by using `anyOf` instead, but that pattern is much
less used.

### `if`/`then`

We can describe this kind of constraint more efficiently and with with better
error messaging by using `if`/`then`. This allows the user to explicitly specify
the constraint to be used to select which alternative the schema should be used
to validate the schema. However, this pattern has problems of it's own. It's
verbose, error prone, and not particularly intuitive, which leads most people to
avoid it.

```jsonschema
{
  "allOf": [
    {
      "if": {
        "properties": {
          "foo": { "const": "aaa" }
        },
        "required": ["foo"]
      },
      "then": { "$ref": "#/$defs/foo-aaa" }
    },
    {
      "if": {
        "properties": {
          "foo": { "const": "bbb" }
        },
        "required": ["foo"]
      },
      "then": { "$ref": "#/$defs/foo-bbb" }
    }
  ]
}
```

> ## [Appendix] Alternatives Considered {#alternatives}

Author: I think the alternatives need to be listed in the same area as the limitations (which arguably need to stay with the solution).  This should be moved back up.
